Make required atlantis/apply status check work with mergeable

rayterrill opened the pull request:
Building on the work of @AndreZiviani, @nishkrishnan, and probably many others - another attempt to try to get mergeable working with required "atlantis/apply" status checks.

First contribution attempt for Atlantis 🙏  - we use this tool heavily and would love to contribute back and get this working if possible - it's a critical pain point in our workflow (and seems similar for others) - would love to partner up with a more experience Atlantis contrib to work through any issues, concerns, etc.

Tests passing locally (via `make test`).

AndreZiviani commented on server/events/vcs/github_client.go:
```diff
@@ -313,6 +347,10 @@ func (g *GithubClient) PullIsMergeable(repo models.Repo, pull models.PullRequest
 	//            hooks. Merging is allowed (green box).
 	// See: https://github.com/octokit/octokit.net/issues/1763
 	if state != "clean" && state != "unstable" && state != "has_hooks" {
+		if state == "blocked" && status && approved.IsApproved {
```
I think the issue here is that state will always be `blocked` (assuming the pending apply status) but you can't know for certain why it is blocked (GitHub api does not specify the reason, only that it is blocked)

I'm not sure if other status and approval are the only things that would block a PR

rayterrill replied:
@AndreZiviani for sure - I anticipated this discussion - that's definitely the majority of the issue here - trying to disentangle what's making up that state.

AndreZiviani commented:
Also you need to actually add the pending apply status [here](https://github.com/runatlantis/atlantis/blob/master/server/events/plan_command_runner.go#L105) like I did [here](https://github.com/runatlantis/atlantis/pull/2053/files#diff-6583ec7260b28e573c74e18e783ee24ba8dce7d0a2e6929c105cc7e74d3d9c6fR107) but keep in mind that this change will affect all VCS not only GitHub

Make sure to read the comments in my [initial PR](https://github.com/runatlantis/atlantis/pull/2053), [this issue](https://github.com/runatlantis/atlantis/issues/2138) and my [last attempt](https://github.com/runatlantis/atlantis/pull/2188)

rayterrill commented:
> Also you need to actually add the pending apply status [here](https://github.com/runatlantis/atlantis/blob/master/server/events/plan_command_runner.go#L105) like I did [here](https://github.com/runatlantis/atlantis/pull/2053/files#diff-6583ec7260b28e573c74e18e783ee24ba8dce7d0a2e6929c105cc7e74d3d9c6fR107) but keep in mind that this change will affect all VCS not only GitHub

@AndreZiviani can you explain a little more about the need for the pending apply status? I saw that, but didn't totally understand the need there. I'll dig into that as well, but hoping you can clarify for me.

AndreZiviani commented:
> [...] another attempt to try to get mergeable working with required "atlantis/apply" status checks.
I assumed that you want to add a pending "atlantis/apply" status check after you run the plan, its been a while since I've looked into this but I don't think it does this by default, if thats the case then you would need to do what I said earlier (this PR as is only does half of the work)

rayterrill commented:
> I assumed that you want to add a pending "atlantis/apply" status check after you run the plan, its been a while since I've looked into this but I don't think it does this by default, if thats the case then you would need to do what I said earlier (this PR as is only does half of the work)

I think I'm still missing something here - I've built this code, and when I run `atlantis plan` or `atlantis apply`, I indeed see the pending updates come in. I guess I'm not understanding why something would need to be added here - just not familiar enough with the codebase. I'm not changing anything about how the "normal" plan/apply routines would run - just adjusting the ruleset used to determine if an apply should be allowed - and only on the github side of things (so ideally no side-effects on any other vcs).

AndreZiviani commented:
when you run `atlantis plan` it automatically adds a pending apply status? it doesn't for me
![image](https://user-images.githubusercontent.com/7469258/183965679-fdc3c9b8-8806-4b9e-8135-e90df8fa0feb.png)

rayterrill commented:
I'm sorry @AndreZiviani I'm still not understanding. When I grab atlantis@master, and then compare it to my PR, the results appear identical to me, which is what I was going for.

When I run `atlantis plan`, I see this:
<img width="924" alt="image" src="https://user-images.githubusercontent.com/4378762/184016181-67d21e08-7aa2-4cdf-bd57-68e73f80ad45.png">

This seems correct to me - the plan is done and successful, and the apply hasn't run yet, but is expected. I wouldn't expect to see a pending apply at this point - because I haven't done anything with the apply stage.

When I run `atlantis apply`, I see the following checks:
<img width="924" alt="image" src="https://user-images.githubusercontent.com/4378762/184016679-a5043309-96e3-4abd-82ec-4a7cb483d8df.png">

This also seems correct to me, and it's the same behavior as with atlantis@master.

AndreZiviani commented:
Then I apologize for the confusion, I misunderstood what you are trying to do.

LGTM

chicocvenancio commented:
I think this will have the same issue as #2311, won't it?

Meaning approvals by people not in the CODEOWNER file will be allowed to apply.

It might be the case I'm misunderstanding the issue in #2112, the code here, or even github's apis, but I think to support the use case of requiring CODEWONER approval we need github's [pullrequest](https://docs.github.com/en/graphql/reference/objects#pullrequest) graphql object with reviewDecision field.

AndreZiviani commented:
You are absolutely right @chicocvenancio , that’s the main reason why I gave up trying to implement something like this

lilincmu commented:
Adding more context to the above discussion regarding `reviewDecision`.

As pointed out in https://github.com/runatlantis/atlantis/issues/2112#issuecomment-1116376830 and its following discussion, we might need to use the `reviewDecision` in GraphQL API as the approval status, instead of using REST API. In short, the approval status from REST API does not take `CODEOWNERS` into consideration, so anyone can approve a PR and then apply it, even when `apply_requirements: [mergeable]` and `CODEOWNERS` are configured. (It's another use case where people want to protect their infra from being applied arbitrarily. They want a PR to be applied, only after it satisfies certain brach protection rules, in this case, approved by code owners, and becomes mergeable.)

rayterrill commented:
@lilincmu Let me test that out - that's definitely a scenario we're using, and I can pretty easily test that out in our environment using a build of this PR. Thank you!

lilincmu reviewed (commented):
+1 on the comment from @AndreZiviani A feature flag might buy us some time to polish the feature without having to revert it immediately due to potential regression.
> This behavior would also need to be enabled/disabled via a feature flag in order to not break existing use-cases like the ones I linked before

rayterrill commented on server/events/vcs/github_client.go:
```diff
@@ -296,12 +297,45 @@ func (g *GithubClient) PullIsApproved(repo models.Repo, pull models.PullRequest)
+	if err != nil {
+		return false, errors.Wrap(err, "getting pull request approval")
+	}
+	fmt.Printf("Approval: %v\n", approved.IsApproved)
```
Yuck this is leftover from me stepping through the code. Will remove.

rayterrill commented on server/events/vcs/github_client.go:
```diff
@@ -296,12 +297,45 @@ func (g *GithubClient) PullIsApproved(repo models.Repo, pull models.PullRequest)
 }
 
+// Checks Statuses for PR, excluding atlantis apply. Returns true if all other statuses are not in failure.
+func (g *GithubClient) GetStatus(repo models.Repo, pull *github.PullRequest) (bool, error) {
```
Updated to "GetCombinedStatusMinusApply"

lilincmu replied:
Sounds better! Thanks.

rayterrill commented on server/events/vcs/github_client.go:
```diff
@@ -296,12 +297,45 @@ func (g *GithubClient) PullIsApproved(repo models.Repo, pull models.PullRequest)
 		return false, errors.Wrap(err, "getting pull request")
 	}
+
+	approved, err := g.PullIsApproved(repo, pull)
```
Updated.

lilincmu replied:
Nice. Thank you!

rayterrill commented:
@lilincmu Added a new GetPullReviewDecision function which calls the graphql api - this indeed takes into account CODEOWENRS and correctly returns "REVIEW_REQUIRED" when a CODEOWNER has not approved the request but someone else has.

Is there an example of featureflag functionality being used already that I could leverage while staying aligned with the codebase?

jamengual commented:
https://github.com/runatlantis/atlantis/pull/2362

rayterrill commented:
@jamengual it doesnt look like atlantis passes userConfig down to NewGithubClient. I could definitely add that in, might be worth adding in the entire userConfig to allow additional flags to modify how that works in the future? Is there another way to do this?

chicocvenancio commented:
Do we still need the feature flag with the current PR? Seems to me this should just work as is currently documented.

rayterrill commented:
It's a good question. I'm a 1st time contributor and thus looking for guidance from more seasoned contribs.

No issue with feature flag rollout to minimize disruptions/accidentally introducing something else - in particular because mergeable is a meta/rollup status - I'm a little concerned we're missing something else other than just approval.

As far as the feature flag though - just not totally sure how to introduce that without having to change a major func/method signature.

lilincmu commented on server/events/vcs/github_client.go:
```diff
@@ -296,12 +297,62 @@ func (g *GithubClient) PullIsApproved(repo models.Repo, pull models.PullRequest)
+	}
+
+	//compile for performance reasons
+	matcher, _ := regexp.Compile("atlantis/apply")
```
We have to pass `vcsstatusname` down here since it's actually configurable. See the [counterpart of GitLab](https://github.com/runatlantis/atlantis/blob/master/server/events/vcs/gitlab_client.go#L216) as an example.

rayterrill replied:
Forgot about that - will fix.

lilincmu commented on server/events/vcs/github_client.go:
```diff
@@ -296,12 +297,62 @@ func (g *GithubClient) PullIsApproved(repo models.Repo, pull models.PullRequest)
+
+	for _, r := range status.Statuses {
+		if !matcher.MatchString(*r.Context) {
+			if *r.State == "failure" {
```
Do we need to return `false` when checks are `pending`?

lilincmu commented:
> it doesnt look like atlantis passes userConfig down to NewGithubClient. I could definitely add that in, might be worth adding in the entire userConfig to allow additional flags to modify how that works in the future? Is there another way to do this?

@rayterrill  Alternatively, you might add the new flag in [`pullReqStatusFetcher`](https://github.com/runatlantis/atlantis/blob/master/server/events/vcs/pull_status_fetcher.go#L12) and pass it down from https://github.com/runatlantis/atlantis/blob/master/server/server.go#L622. Later, the flag can be passed in as a parameter when `PullIsMergeable` calls are made.

> Do we still need the feature flag with the current PR? Seems to me this should just work as is currently documented.

@chicocvenancio Ideally we don't, however, given the history of the multiple attempts for this feature, I think we can be a bit pessimistic on it.

rayterrill commented:
Tracking another issue - looks like a previous PR attempt ran into issues with Github Actions-based checks.

I added one to a test PR, and indeed, the CombinedStatus API doesn't appear to take this into account - those get reported under CheckSuite/CheckRun apis. I'll look at getting that added as well, and update the GetCombinedStatus API to handle both.

rayterrill commented:
@lilincmu might it make more sense to build a new "GithubConfig" struct, and pass that into NewGithubClient? That would give a sustainable, go-forward way to influence github-specific behavior in the future.

Unless I'm misunderstanding - adding it into pullReqStatusFetcher feels a little too heavyhanded - this flag really should only impact the github side of things, would love to keep it constrained to that portion of the code vs modifying what appears to be more global parts of the code.

lilincmu commented:
> might it make more sense to build a new "GithubConfig" struct, and pass that into NewGithubClient? That would give a sustainable, go-forward way to influence github-specific behavior in the future.
> 
> Unless I'm misunderstanding - adding it into pullReqStatusFetcher feels a little too heavyhanded - this flag really should only impact the github side of things, would love to keep it constrained to that portion of the code vs modifying what appears to be more global parts of the code.

@rayterrill Agree. I think the approach you described makes more sense!

rayterrill commented:
- Added feature flag protection for the new functionality with a new GithubConfig struct and CLI flags - default false
- Added check against the check suites API to handle app-based checks (this fixed the issue I had with a GH action-based check)

lilincmu reviewed (approved):
This looks good to me! Thanks for the contribution.

One last thing I can think of is to add a section in [server configuration](https://github.com/lilincmu/atlantis/blob/master/runatlantis.io/docs/server-configuration.md) doc. If you can add it in this PR, it would be wonderful.

rayterrill commented:
I did think an issue this wknd that I'm testing for now - I don't think the checks api has any indication about required checks vs not - looks like the current logic will block if any of the checks fail, not just required ones. I'll add some additional logic to only take into account required status checks.

rayterrill commented:
Added additional logic to suss out only the required check runs and ignore non-required, added server config doc.

rayterrill commented:
@lilincmu do you know if there's a next step to get a maintainer to unlock the additional workflows?

rayterrill commented:
Anything else needed to move this along?

jamengual commented:
we will merge it now and release it on the prerelease @rayterrill

rayterrill commented:
Wow thank you @jamengual ❤️ !

nitrocode commented:
Nice job @rayterrill !

For anyone else who finds this PR from the 0.19.9 release page, remember to set one of these to allow setting atlantis/apply as a requirement. The following is documented by @chroju in PR https://github.com/runatlantis/atlantis/pull/2568.

```
ATLANTIS_GH_MERGEABLE_BYPASS_APPLY=true
```

or flag

```
--gh-allow-mergeable-bypass-apply
```

jamengual commented:
https://github.com/runatlantis/atlantis/pull/2583

nitrocode commented:
@rayterrill @briancurt @minamijoyo apologies for the incorrect env var. I found out the hard way as well. Please make sure to use the correct input. The docs have the correct input flag and name now.

Yasmine92 commented:
```
{"level":"warn","ts":"2022-10-19T13:01:14.604Z","caller":"events/apply_command_runner.go:114","msg":"unable to get pull request status: fetching mergeability status for repo: ***/***, and pull number: 241: getting pull request status: getting required status checks: GET https://api.github.com/repos/***/****/branches/main/protection: 404 Not Found []. Continuing with mergeable and approved assumed false","json":{"repo":"***/***","p
ull":"241"},"stacktrace":"github.com/runatlantis/atlantis/server/events.(*ApplyCommandRunner).Run\n\tgithub.com/runatlantis/atlantis/server/events/apply_command_runner.go:114\ngithub.com/runatlantis/atlantis/server/events.(*DefaultCommandRunner).RunComme
ntCommand\n\tgithub.com/runatlantis/atlantis/server/events/command_runner.go:296"} 
```

alexlo03 commented:
@Yasmine92 Yes to use the feature the Atlantis GH user needs more permissions

alexlo03 commented on server/events/vcs/github_client.go:
```diff
@@ -296,12 +299,118 @@ func (g *GithubClient) PullIsApproved(repo models.Repo, pull models.PullRequest)
+// GetCombinedStatusMinusApply checks Statuses for PR, excluding atlantis apply. Returns true if all other statuses are not in failure.
+func (g *GithubClient) GetCombinedStatusMinusApply(repo models.Repo, pull *github.PullRequest, vcstatusname string) (bool, error) {
+	//check combined status api
+	status, _, err := g.client.Repositories.GetCombinedStatus(g.ctx, repo.Owner, repo.Name, *pull.Head.Ref, nil)
```
I think this might be https://github.com/runatlantis/atlantis/issues/2601 - here the repo.Owner and repo.Name come from the PR Destination repository, and the `pull.Head.Ref` may be the branch name in the Fork (Source repository)

From https://docs.github.com/en/rest/commits/statuses#get-the-combined-status-for-a-specific-reference it's not clear how to do this for commits on a fork.
